pkg/binary/envoy/fetch_test.go

```diff
@@ -17,26 +17,134 @@ package envoy
 import (
 	"fmt"
 	"io"
+	"io/ioutil"
+	"log"
 	"net/http"
 	"net/http/httptest"
 	"os"
 	"path/filepath"
 	"testing"
 
-	"github.com/mholt/archiver/v3"
 	"github.com/stretchr/testify/require"
+	"github.com/ulikunitz/xz"
 
-	reference "github.com/tetratelabs/getenvoy/pkg"
 	"github.com/tetratelabs/getenvoy/pkg/globals"
-	"github.com/tetratelabs/getenvoy/pkg/manifest"
+	"github.com/tetratelabs/getenvoy/pkg/internal"
 	manifesttest "github.com/tetratelabs/getenvoy/pkg/test/manifest"
 	"github.com/tetratelabs/getenvoy/pkg/test/morerequire"
 )
 
 const envoyVersion = "1.17.2" // This is only for unit testing: we don't need to use latest.
+var fakeLogger = log.New(io.Discard, "", 0)
+
+func TestFetchEnvoyError(t *testing.T) {
+	tempDir, removeTempDir := morerequire.RequireNewTempDir(t)
+	envoyPath := filepath.Join(tempDir, "envoy")
+	defer removeTempDir()
+
+	targz := filepath.Join(tempDir, "envoy.tar.gz")
+	require.NoError(t, internal.Tar(targz, tempDir, internal.GzipWriter))
+
+	var realHandler func(w http.ResponseWriter, r *http.Request)
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if realHandler != nil {
+			realHandler(w, r)
+		} else {
+			w.WriteHeader(http.StatusNotFound)
+		}
+	}))
+	defer server.Close()
+
+	t.Run("error on incorrect URL", func(t *testing.T) {
+		e := fetchEnvoy(envoyPath, server.URL, internal.GzipReader)
+		require.EqualError(t, e, fmt.Sprintf(`received 404 status code from %s`, server.URL))
+	})
+
+	realHandler = func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(200)
+	}
+	t.Run("error on empty", func(t *testing.T) {
+		e := fetchEnvoy(envoyPath, server.URL, internal.GzipReader)
+		require.EqualError(t, e, fmt.Sprintf(`not a valid stream %s: EOF`, server.URL))
+	})
+
+	realHandler = func(w http.ResponseWriter, _ *http.Request) {
+		w.Write([]byte("mary had a little lamb"))
+		w.WriteHeader(200)
+	}
+	t.Run("error on not a tar", func(t *testing.T) {
+		e := fetchEnvoy(envoyPath, server.URL, internal.GzipReader)
+		require.EqualError(t, e, fmt.Sprintf(`not a valid stream %s: gzip: invalid header`, server.URL))
+	})
+
+	// Make the handler return the targz
+	realHandler = func(w http.ResponseWriter, _ *http.Request) {
+		f, err := os.Open(targz)
+		require.NoError(t, err)
+		io.Copy(w, f)
+		w.WriteHeader(200)
+	}
+	t.Run("tar missing bin/envoy", func(t *testing.T) {
+		e := fetchEnvoy(envoyPath, server.URL, internal.GzipReader)
+		require.EqualError(t, e, fmt.Sprintf(`bin/envoy not found in %s`, server.URL))
+	})
+}
+
+func TestFetchEnvoy(t *testing.T) {
```

Comment:
ended up having to add more tests as only now I realized we are downloading both xz and gz archives